File: 6 kyu/PI approximation/solution.go
package kata

import (
  "math"
  "fmt"
)

func IterPi(epsilon float64) (int, string) {
  n, v := 1.0, 1.0
  for math.Abs(math.Pi - 4*v) > epsilon {
    v += math.Pow(-1, n)/(2*n + 1)
    n += 1
  }
  return int(n), fmt.Sprintf("%.10f", 4*v)
}
________________________________________
package kata

import (
    "math"
    "fmt"
)

func IterPi(epsilon float64) (int, string) {
    const PI = 3.14159265358979323846
    var divisor float64 = 1.0
    var sign float64 = 1.0
    var count int = 0
    var sum float64 = 0.0
    for ; math.Abs(sum - PI) > epsilon ; {
        sum += sign * 4.0 / divisor
        divisor += 2.0
        sign *= -1.0
        count += 1
    }
    return count, fmt.Sprintf("%.10f", sum)
}
________________________________________
package kata
import ("math";"fmt")
func IterPi(epsilon float64) (int, string) {
  n, p4  := 1, 1.0 
  for i, one:=3.0,-1.0; math.Abs(math.Pi-4*p4)>epsilon; i += 2 {
    p4 += one/i;
    one = -one
    n++
  }
  return n,fmt.Sprintf("%1.10f",4*p4)
}
________________________________________
package kata

import (
  "math"
  "fmt"
  "strconv"
)

func LeibnizTerm(n int) float64 {
  power := math.Pow(-1, float64(n + 1))
  term := 1 / float64(2 * n - 1)
  return power * term
}

func IterPi(epsilon float64) (int, string) {
  i := 1
  result := 0.0
  fmt.Println(math.Pi, i)
  
  for math.Abs(result - math.Pi) > epsilon {
    result += 4 * (LeibnizTerm(i))
    i++
  }
  return i - 1, strconv.FormatFloat(result, 'f', 10, 64)
}
________________________________________
package kata
import "math"
import "fmt"

func IterPi(epsilon float64) (int, string) {
  var pi4 = 1.0
  var sign = -1.0
  var iterations = 1
  var denom = 3.0
  for math.Abs(4.0 * pi4 - math.Pi) >= epsilon {
    pi4 = pi4 + sign * (1.0 / denom)
    sign = sign * (-1.0)
    denom = denom + 2.0
    iterations = iterations + 1
  }
  return iterations, fmt.Sprintf("%.10f", 4.0 * pi4)
}

File: 6 kyu/PI approximation/solution.ex
defmodule Piapprox do

  def iter_pi(epsilon), do: iter_pi(epsilon, 1, 1, 0, 0)
  
  @pi :math.pi
  
  defp iter_pi(epsilon, _num, _den, acc, n) 
    when abs(@pi - 4*acc) <= epsilon, do: [n, trunc10(acc*4)]
  defp iter_pi(epsilon, num, den, acc, n),
    do: iter_pi(epsilon, -num, den + 2, acc + num/den, n + 1)
  
  defp trunc10(x), do: trunc(x*1.0e10)/1.0e10
  
end
________________________________________
defmodule Piapprox do
  # Cache π to save some keystrokes.
  @pi :math.pi()
  
  # The default values for the first iteration.
  def iter_pi(epsilon, pi_4 \\ 0, diff \\ nil, iter \\ 0)
  
  # Returns the current iteration and the calculated value of π when the difference
  # between π/4 and π is smaller than `epsilon`.
  def iter_pi(epsilon, pi_4, diff, iter) when not is_nil(diff) and diff < epsilon do
    [iter, truncate(pi_4 * 4)]
  end
  
  # Calculates the next iteration of π when the difference between π/4 and π is higher
  # than `epsilon`.
  def iter_pi(epsilon, pi_4, _diff, iter) do
    pi_4 = iter_leibniz(pi_4, iter)
    diff = diff(pi_4)
    
    iter_pi(epsilon, pi_4, diff, iter + 1)
  end
  
  # Calculates the `n` iteration of the Leibniz number, provided the given π/4
  # is the value of the `n-1` previous iterations.
  def iter_leibniz(acc_pi_4, n) do
    acc_pi_4 + :math.pow(-1, n) / (2 * n + 1)
  end
  
  # Truncates the value of `n` to 10 digits.
  def truncate(n) do
    trunc(n * :math.pow(10, 10)) / :math.pow(10, 10)
  end
  
  # Calculates the difference between the given π/4 and the actual π.
  def diff(pi_4) do
    abs(pi_4 * 4 - @pi)
  end
end
________________________________________
defmodule Piapprox do

  defp iter_pi_aux(epsilon, cnt, som) do 
    sign = :math.pow(-1, cnt)
    s = som + sign / (2 * cnt + 1)
    r = abs(:math.pi - 4 * s)
    if (r < epsilon) do
      d = (trunc 4 * s * :math.pow(10, 10)) / :math.pow(10, 10)
      [cnt + 1, d]
    else
      iter_pi_aux(epsilon, cnt + 1, s)
    end
  end
  
  def iter_pi(epsilon) do
   iter_pi_aux(epsilon, 0, 0)
  end
  
end
________________________________________
defmodule Piapprox do

  defp iter_pi_aux(epsilon, cnt, som) do 
    sign = :math.pow(-1, cnt)
    s = som + sign / (2 * cnt + 1)
    r = abs(:math.pi - 4 * s)
    if (r < epsilon) do
      d = (trunc 4 * s * :math.pow(10, 10)) / :math.pow(10, 10)
      [cnt + 1, d]
    else
      iter_pi_aux(epsilon, cnt + 1, s)
    end
  end
  
  def iter_pi(epsilon) do
   iter_pi_aux(epsilon, 0, 0)
  end
  
end
________________________________________
defmodule Piapprox do
  @pi :math.pi
  
  def iter_pi(epsilon) do
    upper = @pi + epsilon
    lower = @pi - epsilon
        
    approximate_pi(upper, lower, 0, 1, :plus, 0)
  end
  
  def approximate_pi(upper, lower, candidate, _, _, iteration) when candidate * 4 >= lower and candidate * 4 <= upper do
    [iteration, to_ten_decimal_places(candidate * 4)]
  end
  
  def approximate_pi(upper, lower, candidate, sequence_denom, :minus, iteration) do
    approximate_pi(upper, lower, candidate - (1 / sequence_denom), sequence_denom + 2, :plus, iteration + 1)
  end
  
  def approximate_pi(upper, lower, candidate, sequence_denom, :plus, iteration) do
    approximate_pi(upper, lower, candidate + (1 / sequence_denom), sequence_denom + 2, :minus, iteration + 1)
  end

  def to_ten_decimal_places(n), do: n |> Float.to_string() |> String.slice(0..11) |> String.to_float()
end
